### DISCUSSION TITLE
Any way to apply business logic for deduplication?

### DISCUSSION THREAD

**Discussion by roengram** [ClickHouse/ClickHouse#64772] [Q&A]
Hi, a very simplified table looks like this:
```sql
CREATE TABLE raw (
    dev_id UInt32,
    dev_detail UInt32
) ENGINE = MergeTree()
ORDER BY (did, devdetail);
```

The table sometimes contain very few duplicate rows (exactly same `dev_id` and `dev_detail`). To naively remove duplicates, I can do something like the following:
```sql
SELECT any(dev_id), any(dev_detail) FROM raw1
```

But I'd like to have better deduplication strategy by exploiting business logic. The cardinality of `dev_id` is large, let's say 1 billion. The cardinality of `dev_detail` itself is similarly large, **but for a given `dev_id`, the cardinality of `dev_detail` becomes much smaller, like 100.** I just want to deduplicate rows within the same `dev_id`, not across different `dev_id`.

Will there be any *faster or resource-efficient* way to deduplicate? Changing `raw` table definition - sorting key, having more auxiliarly columns, etc. - is also an option.

**Comment by UnamedRus** [comment]
> just want to deduplicate rows within the same dev_id, not across different dev_id.

It's already working kinda like that:
https://github.com/ClickHouse/ClickHouse/pull/58120
